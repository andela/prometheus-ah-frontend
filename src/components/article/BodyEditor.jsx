import React, { Component } from 'react';
import { Editor } from '@tinymce/tinymce-react';
import PropTypes from 'prop-types';

/**
 * @class BodyEditor
 */
class BodyEditor extends Component {
  /**
   *@returns {*} jsx
   */
  render() {
    const {
      handleChange, value, handleBodyFocusIn, handleBodyFocusOut
    } = this.props;
    return (
      <Editor
        apiKey="zre5x72iyosuc7o3z1vg6lfbkxwdw4b5lr9yrwpgjsnfsca7"
        onEditorChange={handleChange}
        value={value}
        onFocusIn={handleBodyFocusIn}
        onFocusOut={handleBodyFocusOut}
        init={{
          menubar: false,
          theme: 'inlite',
          inline: true,
          mobile: {
            theme: 'mobile',
            plugins: ['autosave', 'lists', 'autolink']
          },
          plugins: [
            'autolink',
            'codesample',
            'link',
            'linkchecker',
            'lists',
            'mediaembed',
            'textcolor',
            'image',
            'emoticons',
          ],
          selection_toolbar:
            `bold italic underline strikethrough|
            h2 h3 | blockquote quicklink | alignleft aligncenter alignright alignjustify |
            forecolor fontsize underlinr`,
          images_reuse_filename: true,
          toolbar: 'emoticons'
        }}
      />
    );
  }
}
BodyEditor.propTypes = {
  handleChange: PropTypes.func,
  value: PropTypes.string,
  handleBodyFocusIn: PropTypes.func,
  handleBodyFocusOut: PropTypes.func,

};

export default BodyEditor;
