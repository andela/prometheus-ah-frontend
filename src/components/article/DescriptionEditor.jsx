import React, { Component } from 'react';
import { Editor } from '@tinymce/tinymce-react';
import PropTypes from 'prop-types';


/**
 * @class BodyEditor
 */
class DescriptionEditor extends Component {
  /**
   *@returns {*} jsx
   */
  render() {
    const {
      handleChange, value, handleTitleFocusIn, handleTitleFocusOut
    } = this.props;
    return (
      <Editor
        apiKey="zre5x72iyosuc7o3z1vg6lfbkxwdw4b5lr9yrwpgjsnfsca7"
        onEditorChange={handleChange}
        value={value}
        onFocusIn={handleTitleFocusIn}
        onFocusOut={handleTitleFocusOut}
        init={{
          menubar: false,
          inline: true,
          plugins: [
            'textcolor',
            'lists',
            'powerpaste',
            'linkchecker',
            'contextmenu',
            'autolink'
          ],
          toolbar: 'undo redo | bold italic underline',
          valid_elements: 'strong,em,span[style],a[href]',
          powerpaste_word_import: 'clean',
          powerpaste_html_import: 'clean',
          valid_styles: {
            '*': 'font-size,font-family,color,text-decoration,text-align'
          },
          content_css: [
            '//fonts.googleapis.com/css?family=Playfair+Display:400,400i,700,700i',
          ]
        }}
      />
    );
  }
}

DescriptionEditor.propTypes = {
  handleChange: PropTypes.func,
  value: PropTypes.string,
  handleTitleFocusIn: PropTypes.func,
  handleTitleFocusOut: PropTypes.func,

};
export default DescriptionEditor;
