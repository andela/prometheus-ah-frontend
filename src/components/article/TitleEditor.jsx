import React, { Component } from 'react';
import { Editor } from '@tinymce/tinymce-react';
import PropTypes from 'prop-types';


/**
 * @class BodyEditor
 */
class TitleEditor extends Component {
  /**
   *@returns {*} jsx
   */
  render() {
    const {
      handleChange, value, handleTitleFocusIn, handleTitleFocusOut,
    } = this.props;
    return (
      <Editor
        apiKey="zre5x72iyosuc7o3z1vg6lfbkxwdw4b5lr9yrwpgjsnfsca7"
        onEditorChange={handleChange}
        value={value}
        onFocusIn={handleTitleFocusIn}
        onFocusOut={handleTitleFocusOut}
        name="title-editor"
        tagName="h1"
        init={{
          menubar: false,
          inline: true,
          plugins: [
            'textcolor',
            'lists',
            'powerpaste',
            'linkchecker',
            'contextmenu',
            'autolink'
          ],
          toolbar: 'undo redo | bold italic underline',
          valid_elements: 'strong,em,span[style],a[href]',
          powerpaste_word_import: 'clean',
          powerpaste_html_import: 'clean',
          valid_styles: {
            '*': 'font-size,font-family,color,text-decoration,text-align'
          },
          content_css: [
            '//fonts.googleapis.com/css?family=Playfair+Display:400,400i,700,700i',
          ]
        }}
      />
    );
  }
}

TitleEditor.propTypes = {
  handleChange: PropTypes.func,
  value: PropTypes.string,
  handleTitleFocusIn: PropTypes.func,
  handleTitleFocusOut: PropTypes.func,
};
export default TitleEditor;
