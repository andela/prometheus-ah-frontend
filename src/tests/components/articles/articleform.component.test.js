import React from 'react';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { shallow } from 'enzyme';
import ConnectedArticleForm, { ArticleForm } from '../../../components/article/ArticleForm';
import BodyEditor from '../../../components/article/BodyEditor';
import TitleEditor from '../../../components/article/TitleEditor';
import DescriptionEditor from '../../../components/article/DescriptionEditor';

const mockStore = configureMockStore([thunk]);
const props = {
  fetchSingleArticle: jest.fn(),
  onSubmit: jest.fn(),
};
const initialState = {
  title: 'Title',
  body: 'new body',
  description: 'Write a short description of your story',
  tag: '',
  tagList: ['tagA'],
};
const store = mockStore(initialState);
describe('Article Form component', () => {
  it('should render Article Form correctly', () => {
    const wrapper = shallow(<ArticleForm />);
    expect(wrapper).toMatchSnapshot();
  });
  it('should set a new state when body changes', () => {
    const wrapper = shallow(<ArticleForm {...props} />);
    const editor = wrapper.find(BodyEditor);
    editor.dive().simulate('EditorChange', 'new value');
    expect(wrapper.state().body).toEqual('new value');
  });
  it('should set a new state when description changes', () => {
    const wrapper = shallow(<ArticleForm {...props} />);
    const description = wrapper.find(DescriptionEditor);
    description.dive().simulate('EditorChange', 'new description');
    expect(wrapper.state().description).toEqual('new description');
  });
  it('should set a new state when title changes', () => {
    const wrapper = shallow(<ArticleForm {...props} />);
    const title = wrapper.find(TitleEditor);
    title.dive().simulate('EditorChange', 'new title');
    expect(wrapper.state().title).toEqual('new title');
  });
  it('should set a new state when title is focused in', () => {
    const wrapper = shallow(<ArticleForm {...props} />);
    const title = wrapper.find(TitleEditor);
    title.dive().simulate('FocusIn');
    expect(wrapper.state().title).toEqual('');
  });
  it('should set a new state when title is focused out', () => {
    const wrapper = shallow(<ArticleForm {...props} />);
    wrapper.setState({ title: '' });
    const title = wrapper.find(TitleEditor);
    title.dive().simulate('FocusOut');
    expect(wrapper.state().title).toEqual('Title');
  });
  it('should set a new state when description is focused in', () => {
    const wrapper = shallow(<ArticleForm {...props} />);
    const description = wrapper.find(DescriptionEditor);
    description.dive().simulate('FocusIn');
    expect(wrapper.state().description).toEqual('');
  });
  it('should set a new state when description is focused out', () => {
    const wrapper = shallow(<ArticleForm {...props} />);
    wrapper.setState({ description: '' });
    const description = wrapper.find(DescriptionEditor);
    description.dive().simulate('FocusOut');
    expect(wrapper.state().description).toEqual('Write a short description of your story');
  });
  it('should set a new state when body is focused in', () => {
    const wrapper = shallow(<ArticleForm {...props} />);
    const body = wrapper.find(BodyEditor);
    body.dive().simulate('FocusIn');
    expect(wrapper.state().body).toEqual('');
  });
  it('should set a new state when body is focused out', () => {
    const wrapper = shallow(<ArticleForm {...props} />);
    wrapper.setState({ body: '' });
    const body = wrapper.find(BodyEditor);
    body.dive().simulate('FocusOut');
    expect(wrapper.state().body).toEqual('Write your story...');
  });
  it('should set a new state onsubmit is passed', () => {
    const wrapper = shallow(<ArticleForm {...props} />);
    const data = {
      title: 'new article',
      body: 'new body',
      description: 'new description',
      tagList: [],
    };
    wrapper.setState(data);
    const submitForm = wrapper.find('form');

    submitForm.simulate('submit', {
      preventDefault: () => {}
    });
    expect(props.onSubmit).toHaveBeenCalledWith(data);
  });
  it('should set a new state when a new tag is entered', () => {
    const wrapper = shallow(<ArticleForm {...props} />);
    const tag = wrapper.find('TagsInput');
    tag.simulate('ChangeInput', 'tag');
    expect(wrapper.state().tag).toEqual('tag');
  });
  it('should set a new state when tag is set in the state', () => {
    const wrapper = shallow(<ArticleForm store={store} {...props} />);
    const tag = wrapper.find('TagsInput');
    tag.simulate('change', 'tag a');
    wrapper.setState({ tagList: ['tag a'] });
    expect(wrapper.state().tagList).toEqual(['tag a']);
  });
  it('should fetch new data when match is passed', () => {
    const props = { // eslint-disable-line
      fetchSingleArticle: jest.fn().mockResolvedValue(Promise.resolve({
        data: {
          article: {
            Tags: [{ id: 1, name: 'writing' }, { id: 2, name: 'reading' }],
            description: 'test',
            body: 'body',
            title: 'title',
          }
        }
      })),
      onSubmit: jest.fn(),
      match: {
        params: {
          slug: 'new-article'
        }
      }
    };
    const wrapper = shallow(<ArticleForm {...props} />);
    wrapper.setState({
      tagList: ['writing', 'reading'],
      description: 'test',
      body: 'body',
      title: 'title',


    });
    expect(props.fetchSingleArticle).toHaveBeenCalled();
    expect(wrapper.state()).toEqual({
      title: 'title',
      body: 'body',
      description: 'test',
      tag: '',
      tagList: ['writing', 'reading'],
    });
  });
  it('should set article to state after create article succeed', () => {
    const wrapper = shallow(<ConnectedArticleForm store={store} {...props} />);
    wrapper.setState({
      title: 'new title',
      body: 'new body',
      description: 'new description',
    });
    expect(wrapper.state()).toEqual({
      title: 'new title',
      body: 'new body',
      description: 'new description',
    });
  });
});
