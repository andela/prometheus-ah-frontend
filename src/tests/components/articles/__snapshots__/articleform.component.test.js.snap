// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Article Form component should render Article Form correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="row"
  >
    <div
      className="col-md-7 align-self-start"
    >
      <TagsInput
        addKeys={
          Array [
            9,
            13,
          ]
        }
        addOnBlur={false}
        addOnPaste={true}
        className="tag-box"
        disabled={false}
        focusedClassName="react-tagsinput--focused"
        inputProps={Object {}}
        inputValue=""
        maxTags={5}
        onChange={[Function]}
        onChangeInput={[Function]}
        onlyUnique={false}
        pasteSplit={[Function]}
        preventSubmit={true}
        removeKeys={
          Array [
            8,
          ]
        }
        renderInput={[Function]}
        renderLayout={[Function]}
        renderTag={[Function]}
        tagDisplayProp={null}
        tagProps={
          Object {
            "className": "react-tagsinput-tag",
            "classNameRemove": "react-tagsinput-remove",
          }
        }
        validationRegex={/\\.\\*/}
        value={Array []}
      />
    </div>
    <div
      className="col-sm-5 col-md-5 submit d-flex justify-content-sm-end justify-content-md-end"
    >
      <input
        className="btn btn-sm btn-primary btn-primary-outline btn-update"
        disabled={true}
        id="publish"
        type="submit"
        value="Publish"
      />
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col-md-12"
    >
      <div
        className="card"
        id="format-title"
      >
        <div
          className="card-title"
        >
          <TitleEditor
            handleChange={[Function]}
            handleTitleFocusIn={[Function]}
            handleTitleFocusOut={[Function]}
            value="Title"
          />
        </div>
      </div>
      <div
        className="description"
      >
        <DescriptionEditor
          handleChange={[Function]}
          handleTitleFocusIn={[Function]}
          handleTitleFocusOut={[Function]}
          value="Write a short description of your story"
        />
      </div>
    </div>
    <div
      className="card editor-body"
    >
      <div
        className="card-body"
      >
        <BodyEditor
          handleBodyFocusIn={[Function]}
          handleBodyFocusOut={[Function]}
          handleChange={[Function]}
          value="Write your story..."
        />
      </div>
    </div>
  </div>
</form>
`;
